ci(docs-infra): test PR previews on CI

gkalpak opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [ ] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->
```
[ ] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] angular.io application / infrastructure changes
[ ] Other... Please describe:
```

## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: N/A


## What is the new behavior?


## Does this PR introduce a breaking change?
```
[ ] Yes
[ ] No
```

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

gkalpak committed:
refactor(docs-infra): fix method name (getPrfromBranch --> getPrFromBranch)

gkalpak committed:
feat(docs-infra): add API endpoint for checking if PR can have preview

There several reasons why PRs cannot have (public) previews:
- The PR did not affect any relevant files (e.g. non-spec files in
  `aio/` or `packages/`).
- The PR cannot be automatically verified as "trusted" (based on its
  author or labels).

Note:
The endpoint does not check whether there currently is a (public)
preview for the specified PR; only whether there can be one.

gkalpak committed:
ci(docs-infra): test PR previews on CI

The deployment of PR previews is triggered by the notification webhook
of the `aio_preview` CircleCI job (which creates and stores the build
artifacts).

This commit adds a new job (`test_aio_preview`), which waits for the
preview to be deployed (for PRs that do have a preview) and then runs
some tests against it (currently only PWA tests).

Fixes #23818

mary-poppins commented:
You can preview 923e9ed at https://pr25671-923e9ed.ngbuilds.io/.

mary-poppins commented:
You can preview 229400f at https://pr25671-229400f.ngbuilds.io/.

mary-poppins commented:
You can preview eeca062 at https://pr25671-eeca062.ngbuilds.io/.

mary-poppins commented:
You can preview b69996b at https://pr25671-b69996b.ngbuilds.io/.

gkalpak commented on aio/scripts/test-preview.js:
```diff
@@ -0,0 +1,122 @@
+// Helpers
+function get(url) {
+  console.log(`GET ${url}`);
+  if (url === previewabilityCheckUrl) return Promise.resolve(JSON.stringify({canHavePreview: true}));
```
Temporarily mock the API that is not yet available on the actual preview server, in order to test the CircleCI job.

petebacondarwin commented on .circleci/config.yml:
```diff
@@ -165,6 +165,23 @@ jobs:
+              source "./scripts/ci/env.sh" print
+              cd aio
+              yarn install --frozen-lockfile --non-interactive
+              xvfb-run --auto-servernum node scripts/test-preview.js $CIRCLE_PR_NUMBER $CIRCLE_SHA1 $AIO_MIN_PWA_SCORE
```
For my sake why do you need `xvfb-run --auto-servernum ` here, rather than just running the node script?

gkalpak replied:
I didn't try without, but I'm pretty sure we need `xvfb` when running apps like Chrome. (And the PWA tests use Chrome.)

petebacondarwin replied:
Oh, it uses Chrome? That makes sense.

petebacondarwin reviewed (approved):
Looks good so far. I would be really tempted to change all that async polling stuff to use Observables :-)

gkalpak commented on aio/aio-builds-setup/dockerbuild/scripts-js/lib/preview-server/preview-server-factory.ts:
```diff
@@ -63,10 +63,36 @@ export class PreviewServerFactory {
+        if (!await buildVerifier.getSignificantFilesChanged(pr, significantFilesRe)) {
+          // Cannot have preview: PR did not touch relevant files: `aio/` or `packages/` (except for spec files).
+          res.send({canHavePreview: false, reason: 'No significant files touched.'});
+          logger.log(`PR:${pr} - Cannot have a preview, because it did not touch any significant files.`);
```
Sounds reasonable. However, we don't use the commit for the check (nor does the caller send it over). Should we retrieve it from GitHub?

petebacondarwin replied:
The SHA is available on the client. Perhaps it could be passed to the request? I just think, looking through the logs, it would be helpful to distinguish different checks?

gkalpak replied:
But the SHA passed by the client would be ignored. The check depends on the current PR state (which may be different than the SHA that triggered the job (if for example the user pushed more commits to the PR while the previous build was still running).

I think it would be overkill to use [PR commits API][1], but we could extract the commit from the [PR files API][2] response (e.g. from `contents_url` - it's the part after `?ref=`):

```json
[
  {
    "sha": "bbcd538c8e72b8c175046e27cc8f907076331401",
    "filename": "file1.txt",
    ...
    "blob_url": "https://github.com/octocat/Hello-World/blob/6dcb09b5b57875f334f61aebed695e2e4193db5e/file1.txt",
    "raw_url": "https://github.com/octocat/Hello-World/raw/6dcb09b5b57875f334f61aebed695e2e4193db5e/file1.txt",
    "contents_url": "https://api.github.com/repos/octocat/Hello-World/contents/file1.txt?ref=6dcb09b5b57875f334f61aebed695e2e4193db5e",
    ...
  }
]
```

[1]: https://developer.github.com/v3/pulls/#list-commits-on-a-pull-request
[2]: https://developer.github.com/v3/pulls/#list-pull-requests-files

petebacondarwin replied:
For logging purposes I am interested in matching up the SHA that triggered the check, rather than the current HEAD SHA in the PR, since that is the SHA that will be used for creating the URL of the preview that is going to be passed to the PWA test, right?

petebacondarwin replied:
I don't feel strongly enough about this to push for it.

gkalpak commented:
> I would be really tempted to change all that async polling stuff to use Observables :-)

Ha ha! It crossed my mind too. But by that time I was almost done and the code is pretty mnimal (for our simple usecase), so I didn't feel like bringing another dependency just for that. (If anything, I would bring in a `fetch` polyfill first 😛)

But I don't mind doing it, if you think it is worth it.

petebacondarwin commented:
No need to change to Observables unless you really wanted to.

gkalpak committed:
fix(docs-infra): correctly check PR files on preview server

According to the docs, the response of GitHub's [PR files API][1]
_"includes a maximum of 300 files"_. This means that if a PR contains
more files, it is possible that not all files are retrieved (which
could, for example, give a false negative for the "significant files
touched" check - not likely but possible).

This commit fixes it by using paginated requests to retrieve all changed
files.

[1]: https://developer.github.com/v3/pulls/#list-pull-requests-files

gkalpak committed:
test(docs-infra): fix preview server unit tests on Windows

Some tests where comparing actual with expected paths, without taking
into account that paths will be different on Windows.

This commit uses `path.resolve()` to convert expected paths to their
OS-specific form.

gkalpak committed:
build(docs-infra): avoid race condition in `aio-builds-setup/` npm scripts

Previously, due to multiple scripts re-building during `yarn dev`
initialization, there could be race conditions that led to errors.

This commit fixes it by ensuring `yarn build` is run once (before
the main `yarn dev` script).

gkalpak committed:
build(docs-infra): replace `concurrently` with `npm-run-all` for preview server dev

`npm-run-all` works just as well, but is better at handling termination on Windows.

gkalpak committed:
build(docs-infra): upgrade preview server dependencies

gkalpak committed:
test(docs-infra): add support for source-maps in preview server tests

gkalpak committed:
test(docs-infra): fix test for preview server's `GithubPullRequests`

gkalpak committed:
build(docs-infra): do not exit preview server `dev` script when `build` fails

mary-poppins commented:
You can preview 3577ac4 at https://pr25671-3577ac4.ngbuilds.io/.

gkalpak committed:
test(docs-infra): fix test for preview server's `BuildCleaner` completing prematurely

gkalpak committed:
refactor(docs-infra): use mockable logger

Related discussion:
https://github.com/angular/angular/pull/23576#discussion_r187925949.

mary-poppins commented:
You can preview d0d8fa8 at https://pr25671-d0d8fa8.ngbuilds.io/.

gkalpak committed:
test(docs-infra): remove unnecessary test helpers

`supertest.Request` extends `Promise` and can be used directly without
"promisifying".

gkalpak committed:
docs(docs-infra): update preview server docs to account for recent changes

Mostly (but not exclusively) a follow-up to #23576.

mary-poppins commented:
You can preview 6a33a95 at https://pr25671-6a33a95.ngbuilds.io/.

mary-poppins commented:
You can preview 06870df at https://pr25671-06870df.ngbuilds.io/.

mary-poppins commented:
You can preview a696563 at https://pr25671-a696563.ngbuilds.io/.

gkalpak commented:
This is not ready for proper review (the only thing pending is discussing https://github.com/angular/angular/pull/25671#discussion_r212828000).
@petebacondarwin, (fwiw) only the last five commits are new (to you). I haven't touched the previous ones (which you had already approved), except for rebasing on master.

gkalpak committed:
fix(docs-infra): fix preview server periodic clean-up

Includes the following fixes:

- Fix cron entry format for clean-up script.
  Crontabs in `/etc` should not have a user field. No idea why it used
  to work before, but it started giving errors recently:
  `/bin/sh: root: not found`.

- Set required env variable in clean-up script. (Broken in cc6f36a9d.)
  This was producing the following error:
  `ERROR: Missing required environment variable 'AIO_CIRCLE_CI_TOKEN'!`

- Use the correct path for downloads to be removed. (Broken in cc6f36a9d.)

gkalpak committed:
fix(docs-infra): use correct parameters for paginated requests to GitHub

As it turns out, in GitHub API paginated requests, page numbering is
1-based. (https://developer.github.com/v3/#pagination)

Starting at page 0 (which returns the first page), results in making the
same request twice and logging incorrect numbers (since the first 100
items are listed twice).

mary-poppins commented:
You can preview 574082d at https://pr25671-574082d.ngbuilds.io/.

gkalpak commented:
I added two moe commits, fixing some clean-up releated bugs.

petebacondarwin commented on aio/aio-builds-setup/docs/overview--security-model.md:
```diff
@@ -40,48 +40,57 @@ part of the CI process and serving them publicly.
-4. If necessary, update the corresponding PR's verification status.
-5. Deploy the artifacts to the corresponding PR's directory.
-6. Prevent overwriting previously deployed artifacts (which ensures that the guarantees established
+1. Receive notification from CircleCI of a completed build.
```
just 'cos github pagination is 1-based, it doesn't mean that everyone has to be :-P

gkalpak replied:
It was just messing with my explicitly listing the steps [below][1] :grin:

[1]: https://github.com/angular/angular/pull/25671/files/ad27caf175b2eb2280e94e96c9f851118880d9e0..13d775bf0b55650a807ec29259e6223974067423#diff-d4b19022f526293b8ed7d5487396e0c9R120

petebacondarwin commented on aio/aio-builds-setup/docs/overview--security-model.md:
```diff
@@ -93,31 +102,32 @@ This section describes how each of the aforementioned sub-tasks is accomplished:
-   Essentially, as long as sub-tasks 1, 2 and 3 can be securely accomplished, it is possible to
-   "project" the trust we have in a team's members through the PR to the build artifacts.
+   The artifacts will be stored to the PR's directory, but will not be publicly accessible unless
+   the PR has been verified. Essentially, as long as sub-tasks 2, 3, 4, 5 can be securely
```
2, 3, 4 and 5

petebacondarwin commented on aio/aio-builds-setup/docs/overview--security-model.md:
```diff
@@ -130,10 +140,10 @@ This section describes how each of the aforementioned sub-tasks is accomplished:
+- Adding the specified label on a PR to mark it as trusted, gives the author full control over the
+  content that is hosted for the specific PR preview (e.g. by pushing more commits to it). The user
+  adding the label is responsible for ensuring that this control is not abused and that the PR is
+  either closed (one way of another) or the access is revoked.
```
Here's a question. If the `aio-preview` label is removed from a PR do we make the previews unavailable?

gkalpak replied:
Yes :smiley:
There's a GitHub webhook that notifies the [/pr-updated][1] endpoint. Basically the same way we amke the previews available when you add the label. (We won't remove the @mary-poppins comment with the links, but they will 404.)

[1]: https://github.com/angular/angular/blob/8f81dba367912d5c905dcd90893475e563bc2eea/aio/aio-builds-setup/dockerbuild/scripts-js/lib/preview-server/preview-server-factory.ts#L128-L146

petebacondarwin reviewed (approved):
still LGTM

mary-poppins commented:
You can preview 240993c at https://pr25671-240993c.ngbuilds.io/.

gkalpak commented on .circleci/config.yml:
```diff
@@ -165,22 +165,20 @@ jobs:
-          name: Wait for preview and run tests (if this is a PR)
+          name: Wait for preview and run tests
           command: |
-            [[ "$CIRCLE_PR_NUMBER" == "" ]] || (
```
Once #26030 is merged, this job (which requires `aio_preview`) will only run on PR builds.

gkalpak commented:
Added one commit to simplify the CircleCI job: https://github.com/angular/angular/pull/25671/commits/5c056284badb0e2e17531d80899e9b3397cdef79
(Once #26030 is merged, the `aio_preview`/`test_aio_preview` jobs will only be run on PR builds.)

mary-poppins commented:
You can preview 5c05628 at https://pr25671-5c05628.ngbuilds.io/.

IgorMinar commented on aio/aio-builds-setup/dockerbuild/scripts-sh/clean-up.sh:
```diff
@@ -2,6 +2,7 @@
 set -eu -o pipefail
 
 # Set up env variables
+export AIO_CIRCLE_CI_TOKEN=UNUSED_CIRCLE_CI_TOKEN
```
do we still need this?

gkalpak replied:
It is still needed in for the preview server. It is not needed for the `clean-up` script (which runs independently).
Due to some refactoring in #23576, even if you need one env var you have to load all of them (in [env-variables.ts][1]), so setting `AIO_CIRCLE_CI_TOKEN` to a dummy value here.

(This only affects the `_TOKEN` variables, which are normally read from files (in contrast to the other env vars which are set in `Dockerfile`.)

[1]: https://github.com/angular/angular/blob/cf095d982dec9fe9bfd66d3df6eb4fe3d4034dd9/aio/aio-builds-setup/dockerbuild/scripts-js/lib/common/env-variables.ts

IgorMinar reviewed (approved):
thanks for working through this!

mary-poppins commented:
You can preview bca3aef at https://pr25671-bca3aef.ngbuilds.io/.

gkalpak commented:
Note to caretaker:
The `test_aio_preview` job is expected to fail, because it requires new code that has not yet been deployed to the preview server.
The code will get automatically deployed once this is merged.

kara commented:
@gkalpak The autosquash is failing in the merge script because there are conflicts in one of your fixup commits 

```
CONFLICT (content): Merge conflict in aio/aio-builds-setup/dockerbuild/scripts-js/test/preview-server/preview-server-factory.spec.ts
error: could not apply 4328fa099... fixup! feat(docs-infra): add API endpoint for checking if PR can have preview
```

Could you try autosquashing and resolving the conflicts on your end, since you have context for the changes?

mary-poppins commented:
You can preview f74dfbd at https://pr25671-f74dfbd.ngbuilds.io/.

gkalpak commented:
Done.
